### PR TITLE
Use Bullet 2.89 as a CMake subproject

### PR DESCRIPTION
**Pull request by mosra**
## Motivation and Context

The main motivation here is to get all Habitat users on the same version of Bullet. Compare to the current setup of having the Hab user install Bullet on her system: conda and package managers currently give different versions of Bullet across platforms.

This PR chooses Bullet 2.89 because that's what most current Hab users are "on" (again, depends on platform). However, another big motivation for this PR is to unblock a version upgrade. We expect to upgrade to 3.06 very soon. 

There are various downsides to no longer using a system install of Bullet: increased build time (building Bullet from source), increased conda distribution size, increased space on disk for the Habitat-sim worktree (all the Bullet source code), etc.

@mosra has included a `USE_SYSTEM_BULLET` build option which lets a user get the old behavior and avoid some downsides. I strongly suggest we don't "officially support" this build option. I.e., if a user posts a github issue about a problem with Habitat and system-installed Bullet, our first and probably only reply should be "switch to the submodule Bullet".

Notes from @mosra 

90% of the work is done in Magnum's [FindBullet.cmake](https://github.com/mosra/magnum-integration/blob/master/modules/FindBullet.cmake) and this only uses it. I added a `USE_SYSTEM_BULLET` option for people like me who don't want to build an extra copy of it every time.

Some caveats:

 * Bullet's buildsystem doesn't correctly express dependencies between static libraries and due to how CMake handles linking to targets, it's *impossible* to link them in correct order. This can be only mitigated with CMake 3.13, but I assume Habitat's minimum CMake version requirement is not that high yet. Another alternative is opening a PR on Bullet's repo and fixing it for them. Because of this, bundled Bullet is built as dynamic libraries.
 * However, Bullet doesn't explicitly export symbols in their shared libraries and Habitat compiles with -fvisibility=hidden, which means there are linker errors unless the hidden visibility is disabled for the Bullet subproject.
 * There's a gigaton of CMake warnings coming out of the Bullet submodule. Unfortunately there isn't anything that could be done to fix those, except for (again) opening a PR on their repo.

## How Has This Been Tested

The builds I tested (@eundersander ):
- bundled bullet and system bullet.
- cmake 3.12 (builds bullet shared libs) and cmake 3.14 (builds bullet static libs)
- Fedora, Ubuntu, and Mac
- with and without --bullet

Cc: @Ram81

### CODE REVIEW

**Comment by mosra** [comment]
@Skylion007 oh, dang, forgot about emscripten. There you have to keep `BUILD_SHARED_LIBS` on `OFF` in `dependencies.cmake` and be sure to use CMake 3.13+.

**Comment by Skylion007** [comment]
@mosra Not sure how to handle that inside of CMake. 

Also habitat's minimum CMake requirement is that for the emscriptem build currently so I don't it's that bad to raise the CMake requirement that much.

**Comment by mosra** [comment]
@Skylion007 Pushed a change that hopefully does it. I don't have the Emscripten build set up locally, so unfortunately I can't test myself. Hope you can :)

File: src/cmake/dependencies.cmake
Comment (erikwijmans) [inline src/cmake/dependencies.cmake:151]: Can you add a version check here and build static if >= 3.13?  I'd like to statically link this for conda as that'll reduce the number of things we need to worry about rpath. The conda-build is just grabbing latest cmake, so no need to worry about versions there.

**Review by erikwijmans — COMMENTED** [review]
On the subject of the conda build, can you remove these lines:

https://github.com/facebookresearch/habitat-sim/blob/master/conda-build/macos_matrix_builder.py#L58
https://github.com/facebookresearch/habitat-sim/blob/master/conda-build/linux_matrix_builder.py#L82

https://github.com/facebookresearch/habitat-sim/blob/master/conda-build/habitat-sim/meta.yaml#L17
https://github.com/facebookresearch/habitat-sim/blob/master/conda-build/habitat-sim/meta.yaml#L66

**Comment by mosra** [comment]
If someone is able to take over and finish this, I'll be happy (@eundersander?). I opened this just to unblock everyone with the build issues, but I should really be focusing on something else ;)

**Comment by erikwijmans** [comment]
> If someone is able to take over and finish this, I'll be happy (@eundersander?). I opened this just to unblock everyone with the build issues, but I should really be focusing on something else ;)

Made my requested changes then :)

File: src/cmake/dependencies.cmake
Comment (Skylion007) [inline src/cmake/dependencies.cmake:154]: Actually, we don't error anymore on CMake < 3.13. Don't we have to add a specific additional require or else people could fail on emscripten silently if build_with_bullet and < 3.13. I'd add an error if ```CORRADE_TARGET_EMSCRIPTEN AND CMAKE_VERSION VERSION_LESS 3.13```

Comment (mosra) [reply src/cmake/dependencies.cmake:154]: Don't worry, I already thought about that -- it's done by FindBullet itself ;) https://github.com/mosra/magnum-integration/blob/a897805489102030da09644f264983944ddbf824/modules/FindBullet.cmake#L73-L75

Comment (Skylion007) [reply src/cmake/dependencies.cmake:154]: Ah, just tested this. Good thinking.

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
Just one thing to note: since we are exporting all the debug symbols from Bullet, this does increase our binary size to 267mb.

**Comment by mosra** [comment]
> 267mb

Hahah. Can you put the `-fvisibility=hidden` workaround into an `if()` so it's done only when building as shared?

**Comment by erikwijmans** [comment]
Good catch there @Skylion007

File: src/cmake/dependencies.cmake
Comment (mosra) [inline src/cmake/dependencies.cmake:158]: wait no, this isn't right :D the `add_subdirectory()` needs to be outside this `if`, plus you forgot half of the comment below

... and then at the end, after `add_subdirectory()` this `if()` again, which resets the CMAKE_CXX_FLAGS back

**Comment by eundersander** [comment]
>If someone is able to take over and finish this, I'll be happy (@eundersander?). 
Thanks for helping with this. I can babysit this PR from here out. I'll do a little testing at least before merging this.

File: src/cmake/dependencies.cmake
Comment (mosra) [inline src/cmake/dependencies.cmake:160]: Come on. The comment lines aren't even in correct order. This goes first.

**Comment by Skylion007** [comment]
Why are we using Bullet 2.89? I know it's confusing but it looks like Erwin is using the v3 tag now...

**Comment by mosra** [comment]
Can you please stop doing unapproved changes? @eundersander wanted to stay on 2.89 for now so that's why this PR pins to the 2.89 tag.

**Comment by Skylion007** [comment]
I'll revert. @eundersander Any specific reason we want to be holden to 2.89 vs the latest version?

**Comment by aclegg3** [comment]
> I'll revert. @eundersander Any specific reason we want to be holden to 2.89 vs the latest version?

2.89 was released in Dec. 2019. Current release of Bullet is [v3.06](https://github.com/bulletphysics/bullet3/releases/tag/3.06). Looks like a couple of nice changes here. E.g. added explicit DynamicType (static, kinematic, dynamic). I think we should consider this release version.

**Comment by eundersander** [comment]
> I'll revert. @eundersander Any specific reason we want to be holden to 2.89 vs the latest version?

I'd like to keep this PR as small as possible. Habitat is currently mostly "on" 2.89, in the sense that 1. Mac package manager gives you 2.89, and 2. Andrew Szot told me he's on 2.89.

Let's wait and see what we break here by making Bullet a submodule. After the dust settles, we can upgrade to 3.06.

**Review by eundersander — APPROVED** [review]
LGTM. I added notes about my testing in the PR description.

**Pull request merged**
